File: mobile/lib/logic/timetable.dart
/// Timetable derivations — ported from src/lib/calculations/timetable.ts.
library;

import '../models/timetable_entry.dart';

class FreePeriod {
  const FreePeriod({required this.periodNo, required this.startTime, required this.endTime});
  final int periodNo;
  final String startTime;
  final String endTime;
}

List<TimetableEntry> entriesForDay(List<TimetableEntry> entries, int dayOfWeek) {
  final today = entries.where((e) => e.active && e.dayOfWeek == dayOfWeek).toList()
    ..sort((a, b) => a.startTime.compareTo(b.startTime));
  return today;
}

/// The class you're in or heading to — the first one that hasn't ended yet.
TimetableEntry? nextEntry(List<TimetableEntry> entries, int dayOfWeek, String nowHm) {
  for (final entry in entriesForDay(entries, dayOfWeek)) {
    if (entry.endTime.compareTo(nowHm) >= 0) return entry;
  }
  return null;
}

/// Periods in the week's grid with no class on [dayOfWeek].
///
/// The portal sends rows only for periods actually taught, so a missing row is
/// a free period. Times are recovered from the same period on another day,
/// since a period keeps its slot all week.
List<FreePeriod> freePeriods(List<TimetableEntry> entries, int dayOfWeek) {
  final timesByPeriod = <int, ({String start, String end})>{};
  for (final entry in entries) {
    final period = entry.periodNo;
    if (period != null && !timesByPeriod.containsKey(period)) {
      timesByPeriod[period] = (start: entry.startTime, end: entry.endTime);
    }
  }

  final busy = entries
      .where((e) => e.active && e.dayOfWeek == dayOfWeek)
      .map((e) => e.periodNo)
      .toSet();

  final free = timesByPeriod.entries
      .where((e) => !busy.contains(e.key))
      .map((e) => FreePeriod(periodNo: e.key, startTime: e.value.start, endTime: e.value.end))
      .toList()
    ..sort((a, b) => a.startTime.compareTo(b.startTime));
  return free;
}

/// A run of consecutive periods of the same subject, shown as one block.
///
/// The portal models a three-hour lab as three separate period rows. Listing
/// them separately is technically faithful and practically useless — a student
/// reading "Technical Hour" three times in a row has to work out for
/// themselves that it's one long session.
class ClassBlock {
  const ClassBlock({required this.entries});

  final List<TimetableEntry> entries;

  TimetableEntry get first => entries.first;
  String get startTime => entries.first.startTime;
  String get endTime => entries.last.endTime;
  int get periods => entries.length;
  bool get isMerged => entries.length > 1;
}

/// Groups a day's classes into blocks, merging neighbours that share a subject.
///
/// "Neighbouring" is by position in the sorted day, not by clock time: the
/// portal's periods butt up against each other but a lunch gap doesn't, and a
/// morning and afternoon session of the same subject should stay separate.
List<ClassBlock> classBlocksForDay(List<TimetableEntry> entries, int dayOfWeek) {
  final day = entriesForDay(entries, dayOfWeek);
  if (day.isEmpty) return [];

  final blocks = <ClassBlock>[];
  var current = <TimetableEntry>[day.first];

  for (final entry in day.skip(1)) {
    final previous = current.last;
    // Same subject *and* actually adjacent — the next period starts when the
    // last one ended, give or take the portal's ten-minute changeover.
    final adjacent = entry.subjectId == previous.subjectId &&
        _minutesBetween(previous.endTime, entry.startTime) <= 15;

    if (adjacent) {
      current.add(entry);
    } else {
      blocks.add(ClassBlock(entries: current));
      current = [entry];
    }
  }
  blocks.add(ClassBlock(entries: current));
  return blocks;
}

int _minutesBetween(String from, String to) {
  int minutes(String hhmm) {
    final parts = hhmm.split(':');
    return (int.tryParse(parts.first) ?? 0) * 60 + (parts.length > 1 ? int.tryParse(parts[1]) ?? 0 : 0);
  }

  return minutes(to) - minutes(from);
}

File: src/lib/calculations/timetable.ts
/**
 * Timetable version selection and change detection (SRS §19, §42, §54, §69).
 * Pure functions over plain data — no Firestore access here (see
 * services/timetable for the query layer that feeds these).
 */
import type { TimetableEntryDiff, TimetableEntryDoc, TimetableVersionDoc } from "@/types/timetable";

/**
 * Selects the single published version whose effective range contains `date`.
 *   published AND effectiveFrom <= date AND (effectiveUntil is null OR effectiveUntil >= date)
 * If several published versions overlap (shouldn't happen, but data can be
 * messy), the most recently published one wins.
 */
export function getActiveTimetableVersion(
  versions: TimetableVersionDoc[],
  date: string,
): TimetableVersionDoc | null {
  const candidates = versions.filter((v) => {
    if (v.status !== "published") return false;
    if (v.effectiveFrom > date) return false;
    if (v.effectiveUntil !== null && v.effectiveUntil < date) return false;
    return true;
  });
  if (candidates.length === 0) return null;
  return candidates.reduce((latest, v) =>
    (v.publishedAt ?? "") > (latest.publishedAt ?? "") ? v : latest,
  );
}

export function getEntriesForDay(
  entries: TimetableEntryDoc[],
  dayOfWeek: number,
): TimetableEntryDoc[] {
  return entries
    .filter((e) => e.active && e.dayOfWeek === dayOfWeek)
    .sort((a, b) => a.startTime.localeCompare(b.startTime));
}

function minutesBetween(from: string, to: string): number {
  const toMinutes = (hhmm: string) => {
    const [h, m] = hhmm.split(":").map(Number);
    return (h || 0) * 60 + (m || 0);
  };
  return toMinutes(to) - toMinutes(from);
}

/**
 * A run of consecutive periods of the same subject, shown as one block.
 *
 * The portal models a three-hour lab as three separate period rows. Listing
 * them separately is technically faithful and practically useless — a student
 * reading "Technical Hour" three times in a row has to work out for
 * themselves that it's one long session. Mirrors mobile's ClassBlock
 * (mobile/lib/logic/timetable.dart), which is where this was first ported to.
 */
export interface ClassBlock {
  entries: TimetableEntryDoc[];
}

export function classBlockPeriods(block: ClassBlock): number {
  return block.entries.length;
}

export function classBlockStartTime(block: ClassBlock): string {
  return block.entries[0].startTime;
}

export function classBlockEndTime(block: ClassBlock): string {
  return block.entries[block.entries.length - 1].endTime;
}

/**
 * Groups a day's classes into blocks, merging neighbours that share a
 * subject and are actually adjacent — the portal's own periods butt up
 * against each other, give or take a ten-minute changeover, but a lunch gap
 * or a second session later in the day doesn't, and a morning and afternoon
 * sitting of the same subject stays separate rather than merging on subject
 * alone.
 */
export function classBlocksForDay(entries: TimetableEntryDoc[], dayOfWeek: number): ClassBlock[] {
  const day = getEntriesForDay(entries, dayOfWeek);
  if (day.length === 0) return [];

  const blocks: ClassBlock[] = [];
  let current: TimetableEntryDoc[] = [day[0]];

  for (const entry of day.slice(1)) {
    const previous = current[current.length - 1];
    const adjacent = entry.subjectId === previous.subjectId && minutesBetween(previous.endTime, entry.startTime) <= 15;

    if (adjacent) {
      current.push(entry);
    } else {
      blocks.push({ entries: current });
      current = [entry];
    }
  }
  blocks.push({ entries: current });
  return blocks;
}

/** Finds the next upcoming (or currently running) entry for "today" from `nowTime` onward. */
export function getNextEntry(
  entries: TimetableEntryDoc[],
  dayOfWeek: number,
  nowTime: string,
): TimetableEntryDoc | null {
  const today = getEntriesForDay(entries, dayOfWeek);
  return today.find((e) => e.endTime >= nowTime) ?? null;
}

export interface FreePeriod {
  periodNo: number;
  startTime: string;
  endTime: string;
}

/**
 * Periods that exist in the week's grid but have no class on `dayOfWeek`.
 *
 * The portal publishes a fixed period grid (1..7 here) and only sends rows for
 * periods that are actually taught, so a missing row *is* a free period. Times
 * are recovered from the same period on other days, since a period keeps its
 * slot all week — which is also why a period never taught on any day can't be
 * reported: nothing tells us when it would have been.
 */
export function getFreePeriods(entries: TimetableEntryDoc[], dayOfWeek: number): FreePeriod[] {
  const timesByPeriod = new Map<number, { startTime: string; endTime: string }>();
  for (const entry of entries) {
    if (entry.periodNo != null && !timesByPeriod.has(entry.periodNo)) {
      timesByPeriod.set(entry.periodNo, { startTime: entry.startTime, endTime: entry.endTime });
    }
  }

  const busy = new Set(
    entries.filter((e) => e.active && e.dayOfWeek === dayOfWeek).map((e) => e.periodNo),
  );

  return [...timesByPeriod.entries()]
    .filter(([periodNo]) => !busy.has(periodNo))
    .map(([periodNo, times]) => ({ periodNo, ...times }))
    .sort((a, b) => a.startTime.localeCompare(b.startTime));
}

/** Free periods for the whole week, keyed by day — for "when can I get work done?". */
export function getWeeklyFreePeriods(entries: TimetableEntryDoc[]): Map<number, FreePeriod[]> {
  const taughtDays = [...new Set(entries.filter((e) => e.active).map((e) => e.dayOfWeek))].sort();
  return new Map(taughtDays.map((day) => [day, getFreePeriods(entries, day)]));
}

const FIELD_LABELS: Record<TimetableEntryDiff["changes"][number]["field"], string> = {
  startTime: "Start Time",
  endTime: "End Time",
  room: "Room",
  facultyId: "Faculty",
  subjectId: "Subject",
};

/**
 * Compares entries between two timetable versions for the same subject slot,
 * matched by (dayOfWeek, subjectId). Produces added/removed/changed summaries
 * for the "Timetable Updated" notification (SRS §22, §23, §54).
 */
export function diffTimetableEntries(
  previous: TimetableEntryDoc[],
  next: TimetableEntryDoc[],
  subjectNameById: Map<string, string>,
): TimetableEntryDiff[] {
  const diffs: TimetableEntryDiff[] = [];
  const key = (e: TimetableEntryDoc) => `${e.dayOfWeek}:${e.subjectId}`;
  const prevByKey = new Map(previous.filter((e) => e.active).map((e) => [key(e), e]));
  const nextByKey = new Map(next.filter((e) => e.active).map((e) => [key(e), e]));

  for (const [k, nextEntry] of nextByKey) {
    const subjectName = subjectNameById.get(nextEntry.subjectId) ?? "Unknown subject";
    const prevEntry = prevByKey.get(k);
    if (!prevEntry) {
      diffs.push({ kind: "added", subjectName, changes: [] });
      continue;
    }
    const changes: TimetableEntryDiff["changes"] = [];
    (Object.keys(FIELD_LABELS) as Array<keyof typeof FIELD_LABELS>).forEach((field) => {
      if (prevEntry[field] !== nextEntry[field]) {
        changes.push({
          field,
          label: FIELD_LABELS[field],
          previous: String(prevEntry[field] ?? "—"),
          next: String(nextEntry[field] ?? "—"),
        });
      }
    });
    if (changes.length > 0) diffs.push({ kind: "changed", subjectName, changes });
  }

  for (const [k, prevEntry] of prevByKey) {
    if (!nextByKey.has(k)) {
      const subjectName = subjectNameById.get(prevEntry.subjectId) ?? "Unknown subject";
      diffs.push({ kind: "removed", subjectName, changes: [] });
    }
  }

  return diffs;
}
